### PR TITLE
feat: rustc_pass_by_value lint attribute

### LINKED ISSUE #76935: Attribute for types that should always be passed by value
Inspired by #76914. Currently, we have a lint in `rustc` for types that should *always* be passed by value. Usually, these types are thin wrappers around references, so there is no benefit to an extra layer of indirection. However, this lint uses a hard-coded list of (diagnostic) names (currently `Ty` and `TyCtxt`). We should use an attribute (e.g. `#[rustc_pass_by_value]`) to control this. Note that `Ty` is a type alias, not a `struct`.

This is worthwhile because there are more widely used types that are merely a wrapper around a reference (e.g. `ty::Predicate`). One could also imagine crates in the ecosystem that also do a lot of interning might take advantage of this.

cc @lcnr

### PR DESCRIPTION
**Pull request by mdibaiee**
Useful for thin wrapper attributes that are best passed as value instead
of reference.

Fixes #76935

### CODE REVIEW

**Comment by mdibaiee** [comment]
My PR is not fully functional yet: For some reason when I try to compile code with `#[pass_by_value]`, I get the error below, even though I have registered it as a builtin:

```
error: cannot find attribute `pass_by_value` in this scope
   --> compiler/rustc_middle/src/ty/mod.rs:469:3
    |
469 | #[pass_by_value]
    |   ^^^^^^^^^^^^^
```

I will try to take another look a bit later, but any help is appreciated. Thank you!

P.S. this PR is subject to change as I'm trying to understand the scope better, see my question here: https://github.com/rust-lang/rust/issues/76935#issuecomment-1007346448

File: compiler/rustc_middle/src/ty/mod.rs [hunk 1]
```diff
@@ -462,6 +462,7 @@ impl<'a, 'tcx> HashStable<StableHashingContext<'a>> for TyS<'tcx> {
 }
 
 #[rustc_diagnostic_item = "Ty"]
+#[pass_by_value]
```
Comment (Urgau) [inline compiler/rustc_middle/src/ty/mod.rs:465]: You're build issue most certainly comes the beta compiler (which is used when bootstrapping the compiler) which obviously doesn't understand the new attribute (you're trying to add it, previous rustc version cannot know about it).

The fix is to not apply the attribute when bootstrapping, like this:
```suggestion
#[cfg_attr(not(bootstrap), pass_by_value)]
```
PS: i'm just a guy who something hack the compiler and the standard library, not a reviewer by any means

Comment (mdibaiee) [reply compiler/rustc_middle/src/ty/mod.rs:465]: @Urgau thanks so much! I had suspected that's the cause but didn't know straight off what would the solution be, will try this! 🙇

File: compiler/rustc_lint/src/pass_by_value.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    ///
+    /// ```rust
+    /// fn returns_result() -> Result<(), ()> {
+    ///     Ok(())
+    /// }
+    ///
+    /// fn main() {
+    ///     returns_result();
+    /// }
+    /// ```
```
Comment (lcnr) [inline compiler/rustc_lint/src/pass_by_value.rs:24]: this example doesn't seem related to this lint :thinking:

**Review by lcnr — COMMENTED** [review]
impl looks good to me

considering that PASS_BY_VALUE is an internal lint i would prefer `#[rustc_pass_by_value]` as the attribute for now.

Before merging this I would like to see some tests, both testing the `Ty`, `TyCtxt` interactions, there are some existing `ui-fulldeps` tests for the old `TY_PASS_BY_REFERENCE` lint. And probably some ordinary `ui` tests using custom types

File: compiler/rustc_lint/src/pass_by_value.rs (hunk 2 above)
Comment (mdibaiee) [reply compiler/rustc_lint/src/pass_by_value.rs:24]: @lcnr ah sorry, it isn't, I haven't finished it up yet, I will write tests and update the example soon

**Comment by mdibaiee** [comment]
I tried to write tests for custom types, but for some reason I didn't get any errors from those custom types in my test file. I am not sure if I understand it correctly, but is that because the lint is internal, and the tests might not be considered rustc internal, and so the lint doesn't apply to them?

I did this in `src/test/ui-fulldeps/internal-lints/rustc_pass_by_value.rs`:

```
#[rustc_pass_by_value]
trait CustomTrait {
    fn test<T: CustomTrait>(
        value: T,
        reference: &T, //~ ERROR passing `CustomTrait` by reference
    );
}

#[rustc_pass_by_value]
trait CustomTraitAlias: CustomTrait {
    fn test<T: CustomTraitAlias>(
        value: T,
        reference: &T, //~ ERROR passing `CustomTraitAlias` by reference
    );
}

#[rustc_pass_by_value]
enum CustomEnum {}

impl CustomEnum {
    fn test(
        value: CustomEnum,
        reference: &CustomEnum, //~ ERROR passing `CustomEnum` by reference
    ) {
    }
}

#[rustc_pass_by_value]
struct CustomStruct {}

#[rustc_pass_by_value]
type CustomAlias<'a> = &'a CustomStruct;

impl CustomStruct {
    fn test(
        value: CustomStruct,
        reference: &CustomStruct, //~ ERROR passing `CustomStruct` by reference
    ) {
    }

    fn test_alias(
        value: CustomAlias,
        reference: &CustomAlias, //~ ERROR passing `CustomAlias` by reference
    ) {
    }
}



————

not found errors (from test file): [
    Error {
        line_num: 71,
        kind: Some(
            Error,
        ),
        msg: "passing `CustomTrait` by reference",
    },
    Error {
        line_num: 79,
        kind: Some(
            Error,
        ),
        msg: "passing `CustomTraitAlias` by reference",
    },
    Error {
        line_num: 89,
        kind: Some(
            Error,
        ),
        msg: "passing `CustomEnum` by reference",
    },
    Error {
        line_num: 103,
        kind: Some(
            Error,
        ),
        msg: "passing `CustomStruct` by reference",
    },
    Error {
        line_num: 109,
        kind: Some(
            Error,
        ),
        msg: "passing `CustomAlias` by reference",
    },
]
```

Is there any way to enable those internal lints in this test file? I'll take another look tomorrow and might find some solution but would appreciate any help in the meantime.

**Comment by mdibaiee** [comment]
Same issue as https://github.com/rust-lang/rust/pull/92646#issuecomment-1008169523 happens when trying to write the example for the lint. The example is compiled in an environment that is not internal hence doesn't work (and I can't even use `rustc_middle::ty:Ty` there).

I just looked at other internal lints and they don't have examples. So I'm going to remove the example from this one as well.

File: compiler/rustc_middle/src/ty/context.rs [hunk 3]
```diff
@@ -961,6 +961,7 @@ pub struct FreeRegionInfo {
 /// [rustc dev guide]: https://rustc-dev-guide.rust-lang.org/ty.html
 #[derive(Copy, Clone)]
 #[rustc_diagnostic_item = "TyCtxt"]
+#[cfg_attr(not(bootstrap), rustc_pass_by_value)]
```
Comment (mdibaiee) [inline compiler/rustc_middle/src/ty/context.rs:964]: This is my first time changing something on the compiler, which makes me have to use this cfg. Do we merge the PR with this in place or is there something I should change here before we merge? When do we clean this up?

Comment (lcnr) [reply compiler/rustc_middle/src/ty/context.rs:964]: yes that's fine, these `cfg_attr` get removed the next time we bump the bootstrap compiler, e.g. in https://github.com/rust-lang/rust/pull/90042/files

**Comment by lcnr** [comment]
for tests, you need both `#![warn(rustc::pass_by_value)]` and potentially `// compile-flags: -Zunstable-options` to actually add the internal lints to the lint storage.

**Comment by mdibaiee** [comment]
@lcnr that flag was already there... I realised that the reason it didn't work was that `rustc_pass_by_value`, as it is implemented now (copy from the previous implementation mostly), requires `rustc_diagnostic_item` attribute on the type as well. Is this justified, or should I change this to not require this attribute?

**Comment by lcnr** [comment]
I'd prefer not requiring that and instead only using the type name. I don't think that requiring `#[rustc_pass_by_value]` and  `rustc_diagnostic_item` adds much value, especially as i can't really think of a case where we don't just want to use the name of the type itself in the diagnostic

File: compiler/rustc_lint/src/pass_by_value.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+}
+
+fn path_for_pass_by_value(cx: &LateContext<'_>, ty: &hir::Ty<'_>) -> Option<String> {
+    if let TyKind::Path(QPath::Resolved(_, path)) = &ty.kind {
+        match path.res {
+            Res::Def(_, def_id) if has_pass_by_value_attr(cx, def_id) => {
+                let name = cx.tcx.item_name(def_id).to_ident_string();
+                return Some(format!("{}{}", name, gen_args(path.segments.last().unwrap())));
```
Comment (lcnr) [inline compiler/rustc_lint/src/pass_by_value.rs:55]: `gen_args` is only correct for types without non default type or const params

Please either fix this in this PR or add a `FIXME` here.

File: compiler/rustc_lint/src/pass_by_value.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            }
+            Res::SelfTy(None, Some((did, _))) => {
+                if let ty::Adt(adt, substs) = cx.tcx.type_of(did).kind() {
+                    if has_pass_by_value_attr(cx, adt.did) {
+                        let name = cx.tcx.item_name(adt.did).to_ident_string();
+                        return Some(format!("{}<{}>", name, substs[0]));
```
Comment (lcnr) [inline compiler/rustc_lint/src/pass_by_value.rs:61]: afaict
```rust
#[rustc_pass_by_value]
struct Foo;

impl Foo {
    fn with_ref(&self) {}
}
```
should ice when accessing `substs[0]`. Please add either a FIXME here or fix this in this PR

**Review by lcnr — COMMENTED** [review]
some final nits, then r=me

File: compiler/rustc_lint/src/pass_by_value.rs (hunk 4 above)
Comment (mdibaiee) [reply compiler/rustc_lint/src/pass_by_value.rs:55]: Can you give an example of a case where this is incorrect? I don't understand what "types without non default type or const params" refers to. Thanks!

Comment (lcnr) [reply compiler/rustc_lint/src/pass_by_value.rs:54]:
```rust
#[rustc_pass_by_value
struct Foo<T, const N: usize>([T; N]);
```
the lint should mention `T` and `N` here, but currently only looks at lifetime params

File: compiler/rustc_lint/src/pass_by_value.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            Res::Def(_, def_id) if cx.tcx.has_attr(def_id, sym::rustc_pass_by_value) => {
+                let name = cx.tcx.item_name(def_id).to_ident_string();
+                return Some(format!("{}{}", name, gen_args(path.segments.last().unwrap())));
+            }
+            Res::SelfTy(None, Some((did, _))) => {
+                if let ty::Adt(adt, substs) = cx.tcx.type_of(did).kind() {
+                    if cx.tcx.has_attr(adt.did, sym::rustc_pass_by_value) {
+                        let name = cx.tcx.item_name(adt.did).to_ident_string();
+                        let param =
+                            substs.first().map(|s| format!("<{}>", s)).unwrap_or("".to_string());
```
Comment (lcnr) [inline compiler/rustc_lint/src/pass_by_value.rs:61]: this also doesn't mention `U` for
```rust
struct Foo<T, U, V = u32>(...);
```
I guess the "right" solution would be to print all `substs` which differ from the parameter defaults,

i.e. when encountering `&Foo<i32, i16, u32>` we should recommend `Foo<i32, i16>`.

Comment (mdibaiee) [reply compiler/rustc_lint/src/pass_by_value.rs]: @lcnr I found a method on `TyCtx` that is apparently made for this purpose, so I'm using that now. Let me know if the output looks good to you.

File: compiler/rustc_lint/src/pass_by_value.rs [hunk 7]
```diff
@@ -70,22 +68,29 @@ fn path_for_pass_by_value(cx: &LateContext<'_>, ty: &hir::Ty<'_>) -> Option<Stri
...
+            .filter_map(|arg| match arg {
+                GenericArg::Lifetime(lt) => Some(lt.name.ident().to_string()),
+                GenericArg::Type(ty) => {
+                    let snippet =
+                        cx.tcx.sess.source_map().span_to_snippet(ty.span).unwrap_or_default();
+                    Some(snippet)
                 }
+                GenericArg::Const(c) => {
+                    let snippet =
+                        cx.tcx.sess.source_map().span_to_snippet(c.span).unwrap_or_default();
```
Comment (lcnr) [inline compiler/rustc_lint/src/pass_by_value.rs:85]:
```suggestion
                        cx.tcx.sess.source_map().span_to_snippet(c.span).unwrap_or_else(|_| String::from("_"));
```
:thinking: don't want empty strings in case the span is broken. Don't know how to even get broken strings here, so i don't think this matters too much :p

**Comment by lcnr** [comment]
r=me after this last nit, thanks for your patience 

@bors delegate+

File: compiler/rustc_lint/src/pass_by_value.rs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    None
+}
+
+fn gen_args(cx: &LateContext<'_>, segment: &PathSegment<'_>) -> String {
+    if let Some(args) = &segment.args {
+        let params = args
+            .args
+            .iter()
+            .filter_map(|arg| match arg {
```
Comment (lcnr) [inline compiler/rustc_lint/src/pass_by_value.rs:76]: use a `map` here and return 'String::from("_")` in the `GenericArg::Infer` case

**Comment by mdibaiee** [comment]
@lcnr thanks a million for your reviews! I will mark this with r=lcnr when CI passes 🚀
